libyul/backends/wasm/BinaryTransform.cpp

```diff
@@ -0,0 +1,553 @@
+
+		index++;
+	}
+
+	return bytes(1, uint8_t(Section::TYPE)) + prefixSize(lebEncode(index) + std::move(result));
+}
+
+bytes BinaryTransform::importSection(
+	vector<FunctionImport> const& _imports
+)
+{
+	bytes result = lebEncode(_imports.size());
+	for (FunctionImport const& import: _imports)
+	{
+		uint8_t importKind = 0; // function
+		result +=
+			encode(import.module) +
+			encode(import.externalName) +
+			bytes(1, importKind) +
+			lebEncode(m_functionTypes[import.internalName]);
+	}
+	return bytes(1, uint8_t(Section::IMPORT)) + prefixSize(std::move(result));
+}
+
+bytes BinaryTransform::functionSection(vector<FunctionDefinition> const& _functions)
+{
+	bytes result = lebEncode(_functions.size());
+	for (auto const& fun: _functions)
+		result += lebEncode(m_functionTypes.at(fun.name));
+	return bytes(1, uint8_t(Section::FUNCTION)) + prefixSize(std::move(result));
+}
+
+bytes BinaryTransform::memorySection()
+{
+	bytes result = lebEncode(1);
+	result.push_back(0); // flags
+	result.push_back(1); // initial
+	return bytes(1, uint8_t(Section::MEMORY)) + prefixSize(std::move(result));
+}
+
+bytes BinaryTransform::globalSection()
+{
+	bytes result = lebEncode(m_globals.size());
+	for (size_t i = 0; i < m_globals.size(); ++i)
+		result +=
+			// mutable i64
+			bytes{uint8_t(ValueType::I64), 1} +
+			opcode(Opcode::I64Const) +
+			lebEncodeSigned(0) +
+			opcode(Opcode::End);
+
+	return bytes(1, uint8_t(Section::GLOBAL)) + prefixSize(std::move(result));
+}
+
+bytes BinaryTransform::exportSection()
+{
+	bytes result = lebEncode(2);
+	result += encode("memory") + bytes(1, uint8_t(Export::Memory)) + lebEncode(0);
+	result += encode("main") + bytes(1, uint8_t(Export::Function)) + lebEncode(m_functions.at("main"));
+	return bytes(1, uint8_t(Section::EXPORT)) + prefixSize(std::move(result));
+}
+
+bytes BinaryTransform::customSection(string const& _name, bytes _data)
+{
+	bytes result = encode(_name) + std::move(_data);
+	return bytes(1, uint8_t(Section::CUSTOM)) + prefixSize(std::move(result));
+}
+
+bytes BinaryTransform::codeSection(vector<wasm::FunctionDefinition> const& _functions)
+{
+	bytes result = lebEncode(_functions.size());
+	for (FunctionDefinition const& fun: _functions)
+		result += (*this)(fun);
+	return bytes(1, uint8_t(Section::CODE)) + prefixSize(std::move(result));
+}
+
+bytes BinaryTransform::visit(vector<Expression> const& _expressions)
+{
+	bytes result;
+	for (auto const& expr: _expressions)
+		result += boost::apply_visitor(*this, expr);
+	return result;
+}
+
+bytes BinaryTransform::visitReversed(vector<Expression> const& _expressions)
+{
+	bytes result;
+	for (auto const& expr: _expressions | boost::adaptors::reversed)
+		result += boost::apply_visitor(*this, expr);
+	return result;
+}
+
+bytes BinaryTransform::prefixSize(bytes _data)
+{
+	size_t size = _data.size();
+	return lebEncode(size) + std::move(_data);
+}
+
+bytes BinaryTransform::encode(std::string const& _name)
```

Comment:
What is this encoding? An UTF8 string?

Reply:
The strings are all ascii.

Reply:
The spec has "strings" and "names". Names are used for imports, etc. and they are UTF8: https://webassembly.github.io/spec/core/binary/values.html#binary-name

Reply:
I'll add an assertion.

Reply:
> 6.3.3  Strings
> Strings denote sequences of bytes that can represent both textual and binary data. They are enclosed in quotation marks and may contain any character other than ASCII43control characters, quotation marks (‘”’), or backslash (‘∖’), except when expressed with an escape sequence.
>
> 6.3.4  Names
> Names are strings denoting a literal character sequence.  A name string must form a valid UTF-8 encoding as defined by Unicode45(Section 2.5) and is interpreted as a string of Unicode scalar values.

"Strings" are used in data segments. "Names" are used everywhere else.